### ISSUE TITLE
AS API gives no way to tell ASes about state resets, so their view of room state can diverge over time.

### ISSUE THREAD

**Issue by ara4n** [matrix-org/matrix-doc#1094]
This is equivalent to https://github.com/matrix-org/synapse/issues/2735 but for AS API rather than /sync.

The proposed solution is to extend the AS API so pushed transactions can be decorated with any state updates which have happened due to resets at the point of the event being sent.

I'm a bit unclear on why this has to be a decoration rather than just inserting copies of the restored events in the right place in the series of transactions (which is basically the approach that https://github.com/matrix-org/synapse/issues/2735 seemed to be converging on)?

**Comment by richvdh** [comment]
I think putting the events in the transaction would be fine tbh.

**Comment by richvdh** [comment]
this isn't just about what we've been calling "state resets".

**Comment by ara4n** [comment]
the decoration idea is quite close to how dendrite’s kafka streams associate state changes with event streams apparently

**Comment by ara4n** [comment]
yup, this isn't state resets caused by bugs or unexpected edge cases in the state resolution algorithm, but instead any situation where state resolution causes state to be altered due to resolving a conflict on another branch. The diagram at https://github.com/matrix-org/synapse/issues/2735#issue-281745736 tries to spell it out.

In other news, i spoke to @erikjohnston today about why this is better as a decoration rather than just including the state events in the transaction series, and the answer was basically:

1. It avoids confusing them with normal state events which have just arrived at the server, and highlights that they are weird magical things.
2. We need semantics to say that a state event has been deleted (e.g. to rollback state which the user set on a branch where after resolution it transpires the user had already been kicked).  The current transaction API doesn't have a way to indicate deleted state, whereas we could make the decoration API handle this.

It feels to me that 1 could be fixed by flagging the event as such, and 2 could be fixed in the AS API for the general case (when we get around to supporting deleting state events).  But if everyone thinks decorations are the way to go, i don't care that much to oppose it.
